### ISSUE TITLE
New algorithm for ranking spanning arborescences

### ISSUE THREAD

**Issue by edxu96** [networkx/networkx#4322]
I'd like to contribute by implementing the first algorithm for ranking spanning arborescences (RANK). It's based on [camerini1980ranking]. According to the paper, given a network, RANK computes K most weighted spanning arborescenceos, rooted at a specified node.

The reason for this algorithm is discussed in the paper:

> Given a combinatorial optimization problem P, it may happen that P is too simple a mathematical model of a real problem, so that having at hand a number of alternative solutions is helpful in choosing the most suitable one in the real environment. It may also happen that it is too difficult to incorporate in P some given set of constraints, while it is an easy matter to solve P: in such cases it can be advisable (or even unavoidable) to rank the solutions until one satisfying the extra constraints is found. 

As mentioned in the paper, formal proofs can be found in [camerini1980k].

---

- [camerini1980ranking] Camerini, P. M., Fratta, L., & Maffioli, F. (1980). Ranking arborescences in O (Km log n) time. European Journal of Operational Research, 4(4), 235-242.
- [camerini1980k] Camerini, P. M., Fratta, L., & Maffioli, F. (1980). The k best spanning arborescences of a network. Networks, 10(2), 91-109.

**Comment by dschult** [comment]
Thanks!

**Comment by edxu96** [comment]
Sorry for the delay. I have made some progress today. Now it is able to:

- Rank spanning arborescences (SAs) in descending order.
- Allow the user to specify any numeric edge attribute for determining optimality.

I still have some tasks to make it more complete and user-friendly:

- [ ] Rank SAs in ascending order.
- [ ] Remove dependencies on `pandas` and `loguru`.
- [ ] Use decorator for accepted graph type.
- [ ] Label edges automatically with `_label` edge attribute instead of `label`, to avoid conflicts.
- [ ] Pass code coverage goal.

Should be able to write a PR before next next Monday.

Also, I have implemented another algorithm to count the number of SAs for a directed graph. Will open another issue later.
